src/app/wallets/pay-on-chain.ts

```diff
@@ -0,0 +1,348 @@
+import { getUsernameFromWalletId } from "@app/accounts"
+import { getCurrentPrice } from "@app/prices"
+import { getUser } from "@app/users"
+import {
+  BTC_NETWORK,
+  getOnChainWalletConfig,
+  ONCHAIN_SCAN_DEPTH_OUTGOING,
+} from "@config/app"
+import { checkedToSats, checkedToTargetConfs, toSats } from "@domain/bitcoin"
+import { PaymentSendStatus } from "@domain/bitcoin/lightning"
+import { checkedToOnChainAddress, TxDecoder } from "@domain/bitcoin/onchain"
+import {
+  InsufficientBalanceError,
+  LessThanDustThresholdError,
+  RebalanceNeededError,
+  SelfPaymentError,
+} from "@domain/errors"
+import { WithdrawalFeeCalculator } from "@domain/wallets"
+import { LedgerService } from "@services/ledger"
+import { OnChainService } from "@services/lnd/onchain-service"
+import { LockService } from "@services/lock"
+import { baseLogger } from "@services/logger"
+import { WalletsRepository } from "@services/mongoose"
+import { NotificationsService } from "@services/notifications"
+
+import {
+  checkAndVerifyTwoFA,
+  checkIntraledgerLimits,
+  checkWithdrawalLimits,
+} from "./check-limit-helpers"
+import { getBalanceForWalletId } from "./get-balance-for-wallet"
+import { getOnChainFeeByWalletId } from "./get-on-chain-fee"
+
+const { dustThreshold } = getOnChainWalletConfig()
+
+export const payOnChainByWalletIdWithTwoFA = async ({
+  senderWalletId,
+  amount,
+  address,
+  targetConfirmations,
+  memo,
+  sendAll,
+  payerUserId,
+  twoFAToken,
+}: PayOnChainByWalletIdWithTwoFAArgs): Promise<PaymentSendStatus | ApplicationError> => {
+  const checkedAmount = sendAll
+    ? await getBalanceForWalletId(senderWalletId)
+    : checkedToSats(amount)
+  if (checkedAmount instanceof Error) return checkedAmount
+
+  const user = await getUser(payerUserId)
+  if (user instanceof Error) return user
+
+  const { twoFA } = user
+
+  const twoFACheck = twoFA?.secret
```

Comment:
I think the call should fail if 2fa is not set (as opposed to continue as if the code was correct). wdyt?

Reply:
I think is better to do it in another PR because we will need to update intraledger and lightning too